packages/core/src/env-set/index.ts

```diff
@@ -23,36 +25,73 @@ export enum MountedApps {
   Welcome = 'welcome',
 }
 
+type MultiTenancyMode = 'domain' | 'env';
+
+const enableMultiTenancyKey = 'ENABLE_MULTI_TENANCY';
+const developmentTenantIdKey = 'DEVELOPMENT_TENANT_ID';
+
 const loadEnvValues = () => {
   const isProduction = getEnv('NODE_ENV') === 'production';
   const isTest = getEnv('NODE_ENV') === 'test';
   const isIntegrationTest = isTrue(getEnv('INTEGRATION_TEST'));
   const isHttpsEnabled = Boolean(process.env.HTTPS_CERT_PATH && process.env.HTTPS_KEY_PATH);
+  const isMultiTenancy = isTrue(getEnv(enableMultiTenancyKey));
   const port = Number(getEnv('PORT', '3001'));
   const localhostUrl = `${isHttpsEnabled ? 'https' : 'http'}://localhost:${port}`;
   const endpoint = getEnv('ENDPOINT', localhostUrl);
   const databaseUrl = tryThat(() => assertEnv('DB_URL'), throwErrorWithDsnMessage);
 
+  const { hostname } = new URL(endpoint);
+  const multiTenancyMode: MultiTenancyMode =
+    isMultiTenancy && !net.isIP(hostname) && hostname !== 'localhost' ? 'domain' : 'env';
+  const developmentTenantId = getEnv(developmentTenantIdKey);
+
+  if (!isMultiTenancy && developmentTenantId) {
+    throw new Error(
+      `Multi-tenancy is disabled but development tenant env \`${developmentTenantIdKey}\` found. Please enable multi-tenancy by setting \`${enableMultiTenancyKey}\` to true.`
+    );
+  }
+
+  if (isMultiTenancy && multiTenancyMode === 'env') {
+    if (isProduction) {
+      throw new Error(
+        `Multi-tenancy is enabled but the endpoint is an IP address: ${endpoint.toString()}.\n\n` +
+          'An endpoint with a valid domain is required for multi-tenancy mode.'
+      );
+    }
+
+    console.warn(
+      '[warn]',
+      `Multi-tenancy is enabled but the endpoint is an IP address: ${endpoint.toString()}.\n\n` +
+        `Logto will use \`${developmentTenantIdKey}\` env (current value: ${developmentTenantId}) for tenant recognition which is not supported in production.`
```

Comment:
nit:
```suggestion
        `Logto is using \`${developmentTenantIdKey}\` env (current value: ${developmentTenantId}) for tenant recognition which is not going to be supported in production.`
```

Reply:
is using + is not supported i think. since "production" means the prod environment, no matter whose

Reply:
I thought most likely the user is still in development phase when they see this, so the `production` is still something not happening at the moment. Thus using the future tense. Therefore they will have to remove the `developmentTenantKey` and also switch to `domain mode` before going to production.

But `is using + is not supported` would probably also work, since it doesn't affect the understanding of the message. I'm OK with both.

Reply:
good